Kernel events race fix

aasthagr opened the pull request:
kernel: events: fix walking the waitq race condition

Fixes race condition for k_event_post_internal() in an
SMP environment while walking the waitq. Uses z_sched_waitq_walk()
to safely walk the waitq by using a sched_spinlock.
It should be noted that since walking the wait queue is an
operation of indeterminant length, there exists the possibility
that the sched_spinlock (which is a highly used and contended-for
lock) may be locked for an indeterminant amount of time. However,
it is expected that few threads will be waiting on any given kernel
event object, which should ameliorate this risk.

kernel: events: Add spin lock/unlock pair to fix race

Adds a spin lock/unlock pair after the waiting thread wakes from
waiting to ensure that the posting thread does not corrupt any data
while processing the list of threads to wake.

Fixes #54317

aasthagr committed:
kernel: events: fix walking the waitq race condition

Fixes race condition for k_event_post_internal() in an
SMP environment while walking the waitq. Uses z_sched_waitq_walk()
to safely walk the waitq by using a sched_spinlock.
It should be noted that since walking the wait queue is an
operation of indeterminant length, there exists the possibility
that the sched_spinlock (which is a highly used and contended-for
lock) may be locked for an indeterminant amount of time. However,
it is expected that few threads will be waiting on any given kernel
event object, which should ameliorate this risk.

Fixes #54317

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

andyross commented on kernel/events.c:
```diff
@@ -269,7 +269,18 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
+	 * On SMP systems, threads in k_event_post_internal()'s processing
+	 * loop may timeout before the list finishes processing. The following
+	 * spin lock/unlock pair prevents those threads from executing further
+	 * until that loop is finished processing.
```
I had trouble when this came by the first time.  The effect of the lock/unlock is to prevent **this thread** from continuing until the thread that holds the lock (if one exists) releases it.  I don't think that's consistent with the comment here, which seems to imply that this is going to block other threads, which it can't,... except for the few cycles it holds the lock, during which it's not taking any externally-visible action anyway, so it might as well not be happening from the perspective of other threads.  So... why must **this** thread block, and what might happen if it doesn't?

Can we get a sequence diagram of the synchronization here?  It still feels off to me.

andyross replied:
> The effect of the lock/unlock is to prevent this thread from continuing until the thread that holds the lock (if one exists) releases it.

And to be clear: ...which is a very different state than "no other thread now holds the lock" or "no other thread was able to take the lock and release it a second/third/etc... time"!

aasthagr replied:
@andyross So k_event_wait_internal() waits for the events to be posted. Now let's say core 1 is executing the k_event_post_internal() do/while loop to unpend the threads while walking the linked list.  Now, if a thread on core 2 that was waiting for an event gets timed out while core 1 was in the mid of executing the do/while loop, then thread 2 would be removed from the waitq and will be unpended. Following that sequence thread 2 now would continue execution while we are still walking the linked list in the do/while loop (k_event_post_internal). This could lead to corruption of thread->next_event_link and thread->events. We want to avoid this behavior in SMP systems, hence I say "prevents those threads from executing". Let me know if is confusing from the reader's perspective.

Tried to answer your question. Let me know if it doesn't make sense.

andyross replied:
> Following that sequence thread 2 now would continue execution while we are still walking the linked list in the do/while loop

But how does this code stop thread 2?  Even if it starts spinning on the lock taken here, it will exit that spin loop instantaneously when we release the lock.  You "stop" a thread from progressing by holding a lock, not just taking it.

Now, there **are** synchronization paradigms that can work purely with ordered events.  I'm not saying this can't possibly be fixing a bug.  I'm just saying that if we want to commit a novel lockless algorithm we need to do a lot more work to explain things.

Mostly I just don't buy that this is fixing anything.   Can't we rework this so that it uses traditional synchronization paradigms?

peter-mitsis replied:
I think I see the core of the misunderstanding, and it stems from description in the comment--the spin lock barrier is meant to prevent the current thread from executing.

In the following scenario, thread T1 on core 1 is processing the list of threads to wake up in k_event_post_internal(). Thread T2 is one of the threads in that list. The intent of the spin lock barrier in k_event_wait_internal() is to prevent thread T2 from executing further (on a different core) if it had timed out while T1 is processing that list. As T1 is holding the spin lock, T2 will wait until that spin lock has been released before (safely) continuing.

What are the worst case scenarios should T2 progress before T1 finishes? T2 could wait on another event object, and then when T1 gets around to processing the T2 node in its list, T1 may change the set of events that T2 is now waiting on. Or, T2 could now be part of a different chain of threads being handled by T3, and when T2 is processed by T1, T1 will continue processing T3's chain.

Such scenarios can occur on SMP because the timeout queue and the event's wait queue are protected by spin locks other than the event's spin lock. It does not occur on UP because there are no other cores in play, and the event's spin lock degenerates into an interrupt lock which is sufficient for that case.

Is that spin lock barrier elegant? No. Does it work? I think it does.

andyross replied:
I'm still confused.  Is T1 the thread taking the lock here or the one running on a different cpu?  Could you write this up in event list form (i.e. "Thread X on core A reaches line M; Thread Y on core B is at line N; ..."), explain the broken part in detail (i.e. "X may read Q in one of two different states depending on the race with Y") and then add that to the comment?

But to repeat: the only (!) reliable effect of this code here is to cause k_event_wait_internal() to wait until someone else releases the lock.  So the explanation of the fix should have the form "Once k_spin_lock() returns, we are guaranteed that <insert safe state here>"

peter-mitsis replied:
1. T2 waits on event object A. Events for which it is waiting are identified in T2’s events field.
2. T1 (on core 1) takes the event object A’s spin lock.
3. T1 creates a list of threads to wake linked together via each thread’s next_event_link field.
4. Core 2 processes a timer interrupt. This removes T2 from event object A’s wait queue.
5. T2 wakes and runs on core 2. It is now in an unpredictable state because it is running.
6. Meanwhile on core 1, T1 finally gets to processing T2 in its list of threads to process (created in step 3).
Here comes the meat of the problem ...
7. T1 (which still has event object A’s spin lock) currently updates and references T2’s fields. However, this is unsafe to do so because T2 is not in a known state. Since T2 previously executed, it could be waiting on a different event object (making it more unsafe for T1 to modify/use T2’s events or next_event_link fields).

We can patch this hole by tweaking step 5 above to become ....
5. T2 wakes and runs on core 2. However, before it can do anything of significance, we have it busy wait on event object A’s spin lock.
6. Meanwhile on core 1, T1 finally gets to processing T2 in its list of threads to process (created in step 3).
7. T1 can now safely use T2’s events and next_event_link fields as T1 has the spin lock and T2 is waiting for it.
8. When T1 finishes processing the list it releases the spin lock.
9. T2 is now able to obtain the spin lock and continue.
10. T2 immediately releases it because now it is the only thread that might be updating its events and next_event_link fields.

andyross commented on include/zephyr/kernel/thread.h:
```diff
@@ -264,6 +264,9 @@ struct k_thread {
 
```
Nitpick: can you break the second patch up into one that introduces the new scheduler API and another that uses it in the IPC layer?

aasthagr replied:
@andyross addressed this :) Thank you

andyross commented on kernel/include/ksched.h:
```diff
@@ -64,6 +64,7 @@ void z_ready_thread(struct k_thread *thread);
 void z_requeue_current(struct k_thread *curr);
 struct k_thread *z_swap_next_thread(void);
 void z_thread_abort(struct k_thread *thread);
+void z_sched_wake_thread(struct k_thread *thread, bool is_timeout);
```
This looks like a good addition to the "official scheduler internal API" layer.  Can you move the declaration to the bottom of the file, near z_sched_wake() and write a proper documentation block for it?

andyross commented on include/zephyr/kernel/thread.h:
```diff
@@ -264,6 +264,9 @@ struct k_thread {
 	uint32_t   event_options;
+
+	/** true if timeout must be prevented */
+	bool prevent_timeout;
```
Can we name this something else?  I get that it "prevents handling of thread timeouts", but that relies on the reader understanding that a thread timeout is used exclusively within the kernel to wake up pended threads.  Maybe "no_wake", "no_timeout_unpend" or something similar?

aasthagr replied:
renamed prevent_timeout to no_wake_on_timeout now. Its a longer name but I think it justifies the potential confusion that you pointed out.

andyross commented on kernel/sched.c:
```diff
@@ -794,26 +794,48 @@ ALWAYS_INLINE void z_unpend_thread_no_timeout(struct k_thread *thread)
+#ifdef CONFIG_EVENTS
+		/* Conditionally clear prevent_timeout */
+		thread->prevent_timeout &= is_timeout;
+		prevent = thread->prevent_timeout;
```
Having the reset to prevent_timeout happen implicitly seems a little confusing.  Why must this happen here, and only once?  Why not reset it when the next timeout gets set on the thread instead?

aasthagr replied:
thread->prevent_timeout value gets cleared up here if this routine is called from the context of k_events_post_internal. Its value is set to true when we are in the middle of processing k_events_post_internal and do not want those threads to timeout and execute in the background.  
So we clear this up here for every thread in waitq which was waiting on events, so we can actually continue with timeouts (line 816) after walking the list of threads is complete. It's happening implicitly here just coz we call this routine from both contexts: 1) An actual timeout and 2) preventing a timeout.

andyross commented on kernel/events.c:
```diff
@@ -98,11 +98,19 @@ static int event_walk_op(struct k_thread *thread, void *data)
+		 * will be woken up by k_event_post_internal once they
+		 * have been processed.
+		 */
+		thread->prevent_timeout = true;
```
Dumb question: can't you just use z_abort_timeout() here?  Do you need to re-enable the pre-existing timeout for some reason?  Doesn't look like it at first glance.  If so that would simplify things enormously.

peter-mitsis replied:
What a great idea! I think it would work in most cases, but still leave one small hole to patch.

1. Core 0 starts walking the waitq in k_sched_post_internal (sched_spinlock is held).
2. Core 1 enters sys_clock_announce() and calls the associated timeout handler--z_thread_timeout().
3. Since Core 0 holds sched_spinlock, core 1 spins waiting for it.
4. Core 0 could abort almost all the relevant timeouts, save for the one whose timeout handler is already in progress on Core 1.
5. However, it should be possible to have an early bail out condition in z_thread_timeout() to cover that small hole.

andyross reviewed (commented):
OK, this version I understand much better.  A few nitpicks, but nothing seems scary.

aasthagr committed:
kernel: Add z_sched_wake_thread API

This API wakes up a given thread and is also called from
z_thread_timeout()

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr committed:
kernel: events: fix waitq timeout race condition

Updates events to prevent a timeout from corrupting the list of
threads that needs to be waken up.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

andyross reviewed (approved):
Looks great, thanks
